sqlglot/transforms.py

```diff
@@ -244,9 +258,28 @@ def preprocess(
     """
 
     def _to_sql(self, expression: exp.Expression) -> str:
+        expression_type = type(expression)
+
         expression = transforms[0](expression.copy())
         for t in transforms[1:]:
             expression = t(expression)
-        return getattr(self, expression.key + "_sql")(expression)
+
+        _sql_handler = getattr(self, expression.key + "_sql", None)
+        if _sql_handler:
+            return _sql_handler(expression)
+
+        transforms_handler = self.TRANSFORMS.get(type(expression))
+        if transforms_handler:
+            # Ensures we don't enter an infinite loop. This can happen when the original expression
+            # has the same type as the final expression and there's no _sql method available for it,
+            # because then it'd re-enter _to_sql.
+            if expression_type is type(expression):
+                raise ValueError(
+                    f"Expression type {expression.__class__.__name__} requires a _sql method in order to be transformed."
+                )
+
+            return transforms_handler(self, expression)
+
+        raise ValueError(f"Unsupported expression type {expression.__class__.__name__}.")
```

Comment:
The motivation behind these changes was that we don't have a `_sql` method for `exp.ApproxQuantile`, so the transformation I added would fail at `getattr`. I think it makes `_to_sql` a bit more robust.